### PR TITLE
Prevent runtime prop metadata retrieval when [JsonIgnore] is used

### LINKED ISSUE #59936: [System.Text.Json] Regression: [JsonIgnore] Attribute not Applied to Ref Returns
### Description

Found a regression in `System.Text.Json` where using properties with `ref returns` cannot be ignored for serialization, leading to an unavoidable exception.

Regression was introduced in `System.Text.Json 6.0.0-rc.1.21406.5`, going by the available packages in the `dotnet6` NuGet repo. This package was pushed on Friday, August 6th, 2021.

As I'm not familiar with how dotnet is exactly built, I do not know how to pinpoint the sdk build number to the exact commit causing the issue however; but the build version should hopefully be useful.

Minimal Reproduction:
```csharp
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Serialization;

var sampleText = new TypeWithRefString();
var json = JsonSerializer.Serialize(sampleText); // Exception Here
var deserialized = JsonSerializer.Deserialize<TypeWithRefString>(json); // And Here

internal class TypeWithRefString
{
    [JsonIgnore]
    public ref string NameRef => ref Name;
    public string Name = "How many Richard Landers does it take to write a blog entry? Just one, you rock!";
}
```

Debug in your own desired way. 
This snippet can be pasted and ran over a blank .NET 6 console app. 

### Configuration

N/A

### Regression?

Yes, this use case worked with any version of System.Text.Json <= `6.0.0-rc.1.21406.4`;

### Other information

Exception:
```csharp
System.ArgumentException: 'The type 'System.String&' may not be used as a type argument.'
   at System.RuntimeType.ThrowIfTypeNeverValidGenericArgument(RuntimeType type)
   at System.RuntimeType.SanityCheckGenericArguments(RuntimeType[] genericArguments, RuntimeType[] genericParamters)
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   at System.Text.Json.Serialization.Converters.ObjectConverterFactory.CreateConverter(Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverterFactory.GetConverterInternal(Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetConverterInternal(Type typeToConvert)
   at System.Text.Json.JsonSerializerOptions.DetermineConverter(Type parentClassType, Type runtimePropertyType, MemberInfo memberInfo)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.GetConverter(Type type, Type parentClassType, MemberInfo memberInfo, Type& runtimeType, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.AddProperty(MemberInfo memberInfo, Type memberType, Type parentClassType, Boolean isVirtual, Nullable`1 parentTypeNumberHandling, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.CacheMember(Type declaringType, Type memberType, MemberInfo memberInfo, Boolean isVirtual, Nullable`1 typeNumberHandling, Boolean& propertyOrderSpecified, Dictionary`2& ignoredMembers)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo..ctor(Type type, JsonConverter converter, Type runtimeType, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo..ctor(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.<RootBuiltInConvertersAndTypeInfoCreator>g__CreateJsonTypeInfo|107_0(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetClassFromContextOrCreate(Type type)
   at System.Text.Json.JsonSerializerOptions.GetOrAddClass(Type type)
   at System.Text.Json.JsonSerializerOptions.GetOrAddClassForRootType(Type type)
   at System.Text.Json.JsonSerializer.GetTypeInfo(Type runtimeType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Write[TValue](TValue& value, Type runtimeType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Serialize[TValue](TValue value, JsonSerializerOptions options)
   at Program.<Main>$(String[] args) in C:\Users\sewer\Desktop\Projects\SystemTextJsonBug\SystemTextJsonBug\SystemTextJsonBug\Program.cs:line 12
```

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/59936, a 6.0 regression. Contributes to https://github.com/dotnet/runtime/issues/59364.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 1]
```diff
@@ -797,23 +797,28 @@ private string GenerateCtorParamMetadataInitFunc(TypeGenerationSpec typeGenerati
...
 
                 for (int i = 0; i < paramCount; i++)
                 {
                     ParameterInfo reflectionInfo = parameters[i].ParameterInfo;
 
+                    string parameterTypeRef = reflectionInfo.ParameterType.GetCompilableName();
+
+                    object? defaultValue = reflectionInfo.GetDefaultValue();
+                    string defaultValueAsStr = defaultValue == null
+                        ? $"default({parameterTypeRef})"
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:810]: The `GetParamDefaultValueAsString` already seems to contain handling logic for null "default values". Why do we need to special case that here?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 2]
```diff
@@ -797,23 +797,28 @@ private string GenerateCtorParamMetadataInitFunc(TypeGenerationSpec typeGenerati
...
     {{
         Name = ""{reflectionInfo.Name!}"",
-        ParameterType = typeof({reflectionInfo.ParameterType.GetCompilableName()}),
+        ParameterType = typeof({parameterTypeRef}),
         Position = {reflectionInfo.Position},
         HasDefaultValue = {ToCSharpKeyword(reflectionInfo.HasDefaultValue)},
-        DefaultValue = {GetParamDefaultValueAsString(reflectionInfo.DefaultValue)}
+        DefaultValue = {defaultValueAsStr}
     }};
-    {parametersVarName}[{i}] = {InfoVarName};
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:816]: Is the `InfoVarName` refactoring necessary? Given we're backporting this to GA branch I would try to keep the diff to the essential changes. We can improve in `main` in future iterations.

File: src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs [hunk 3]
```diff
@@ -2056,7 +2056,7 @@ public class Point_MembersHave_JsonIgnore : ITestClass
         public void Verify()
         {
             Assert.Equal(0, X);
-            Assert.Equal(0, Y); // We don't set parameter default value here.
+            Assert.Equal(5, Y);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs:2059]: So it seems like some of the changes here are meant to address `defaultValue` handling in the case of `JsonIgnoreAttribute`? Is this the expected behavior when `Y` doesn't have `JsonIgnore` and we deserialize `{ "X" : 1 }`?

It's clearly something we should be fixing but I think it's required to fix the bug reported in #59936. I would recommend splitting this PR into one that applies a minimal fix for #599396 for GA backporting and a separate one that addresses default value handling in 7.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:810]: I've moved the special casing into `GetParamDefaultValueAsString`. Special casing is needed to provide a `default` value for value-type params and avoid `InvalidCastException` from `null` to those types in the generated ctor invocation logic.

File: src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs:2059]: > Is this the expected behavior when Y doesn't have JsonIgnore and we deserialize { "X" : 1 }?

Yup. Created an issue to track the correct default value handling for 7.0 and reversed related changes - https://github.com/dotnet/runtime/issues/60082.

**Review by steveharter — APPROVED** [review]
LGTM assuming green CI

**Review by eiriktsarpalis — APPROVED** [review]

**Pull request merged**